examples/pytorch/adaptive_sampling/adaptive_sampling.py

```diff
@@ -0,0 +1,475 @@
+        # are ordered the same as curr_frontier.
+
+        is_sparse = False
+        sample_weights = self.sample_weights
+        layer_size = self.layer_nodes[layer_index]
+        current_layer_feature = self.node_feature[curr_frontier]
+
+        src, des, eid = self.graph.in_edges(curr_frontier, form='all')
+        neighbor_nodes = torch.unique(torch.cat((curr_frontier, src), dim=0), sorted=False)
+        sparse_adj = self.norm_adj[curr_frontier, :][:, neighbor_nodes]
+        square_adj = sparse_adj.multiply(sparse_adj).sum(0)
+        if is_sparse:
+            sparse_adj = sparse_adj.tocoo()
+            tensor_adj = torch.sparse.DoubleTensor(sparse_adj.row, sparse_adj.col, sparse_adj.data)
+        else:
+            tensor_adj = torch.tensor(square_adj.A[0])
+        ##compute sampling probability for next layer which is decided by :
+        # 1. attention part derived from node hidden feature
+        # 2. adjacency part derived from graph structure
+        hu = torch.matmul(self.node_feature[neighbor_nodes], sample_weights[:, 0])
+        hv = torch.sum(torch.matmul(self.node_feature[curr_frontier], sample_weights[:, 1]))
+        adj_part = torch.sqrt(tensor_adj)
+        attention_part = F.relu(hv + hu) + 1
+        gu = F.relu(hu) + 1
+        probas = adj_part * attention_part * gu
+        probas = probas / torch.sum(probas)
+        ##build graph between canidates and curr_frontier
+        canidates = neighbor_nodes[probas.multinomial(num_samples=layer_size, replacement=True)]
+        ivmap = {x: i for i, x in enumerate(neighbor_nodes.numpy())}
+        # use matrix operation in pytorch to avoid for-loop
+        curr_padding = curr_frontier.repeat_interleave(len(canidates))
+        cand_padding = canidates.repeat(len(curr_frontier))
+        ##the edges between canidates and curr_frontier composed of
+        # 1. edges in orginal graph
+        # 2. edges between same node(self-loop)
+        has_loops = curr_padding == cand_padding
+        has_edges = self.graph.has_edges_between(cand_padding, curr_padding)
+        loops_or_edges = (has_edges.bool() + has_loops).int()
+        # get neighbor_nodes and corresponding sampling probability
+        num_neighbors = loops_or_edges.reshape((len(curr_frontier), -1)).sum(1)
+        sample_neighbor = cand_padding[loops_or_edges.bool()]
+        q_prob = probas[[ivmap[i] for i in sample_neighbor.numpy()]]
+
+        # get the edge mapping ,-1 if the edge dostn't exist
+        eids = torch.zeros(torch.sum(num_neighbors), dtype=torch.int64) - 1
+        has_edge_ids = torch.where(has_edges)[0]
+        all_ids = torch.where(loops_or_edges)[0]
+        edges_ids_map = torch.where(has_edge_ids[:, None] == all_ids[None, :])[1]
+        eids[edges_ids_map] = self.graph.edge_ids(cand_padding, curr_padding)
+
+        return sample_neighbor, eids, num_neighbors, q_prob
+
+
+class SAGEConv2(nn.Module):
```

Comment:
Change the name `SAGEConv2` to something else (maybe `AdaptiveSAGEConv`?) because it is not GraphSAGE anymore.